File: Cribbage/Utils/Card.cpp
#include "Card.h"

#include <array>
#include <string>

#include "Deck.h"

Card::Card() :
      mSuit(Suit::NoSuit)
    , mName("None")
    , mValue(0)
{}

Card::Card(Suit suit, std::string name, int val) :
      mSuit(suit)
    , mName(name)
    , mValue(val)
{
    if (mValue < 1 || mValue > 10)
    {
        std::cerr << "Error -- value too large: " << mValue << std::endl;
    }
}

Card& Card::operator=(const Card& card)
{
    mSuit = card.mSuit;
    mName = card.mName;
    mValue = card.mValue;
    return *this;
}

bool Card::operator==(const Card& oth) const
{
    return (mSuit == oth.mSuit) && (mName == oth.mName);
}

bool Card::operator!=(const Card& oth) const
{
    return !(*this == oth);
}

bool Card::operator<(const Card& oth) const
{
    int i = Card::getNameIndex(mName);
    int j = Card::getNameIndex(oth.mName);

    if (i < j) { return true; }
    else { return false; }
}

std::ostream& operator<<(std::ostream &os, const Card& c)
{
    std::string suit;
    if (c.mSuit == Suit::NoSuit)   { suit = "None    "; }
    if (c.mSuit == Suit::Spades)   { suit = "spades  "; }
    if (c.mSuit == Suit::Diamonds) { suit = "diamonds"; }
    if (c.mSuit == Suit::Clubs)    { suit = "clubs   "; }
    if (c.mSuit == Suit::Hearts)   { suit = "hearts  "; }

    std::string nameSpaces;
    if (c.mName == "ace")  { nameSpaces += "  "; }
    if (c.mName == "two")  { nameSpaces += "  "; }
    if (c.mName == "four") { nameSpaces += " ";  }
    if (c.mName == "five") { nameSpaces += " ";  }
    if (c.mName == "six")  { nameSpaces += "  "; }
    if (c.mName == "nine") { nameSpaces += " ";  }
    if (c.mName == "ten")  { nameSpaces += "  "; }
    if (c.mName == "jack") { nameSpaces += " ";  }
    if (c.mName == "king") { nameSpaces += " ";  }

    os << suit << "  " << c.mName << nameSpaces; // << " val: " << c.mValue;
    return os;
}

int Card::getNameIndex(std::string name)
{
    for (int n = 0; n < Deck::sNames.size(); n++)
    {
        if (Deck::sNames[n] == name)
        {
            return n;
        }
    }
    return -1;
}

bool Card::contains(const std::vector<Card>& vec, const Card& card)
{
    for (int v = 0; v < vec.size(); v++)
    {
        if (vec[v] == card) { return true; }
    }
    return false;
}

void Card::erase(std::vector<Card>& vec, const Card& card)
{
    int i = -1;
    bool stop = false;
    for (int v = 0; v < vec.size() && !stop; v++)
    {
        if (vec[v] == card) { i = v; stop = true; }
    }
    vec.erase(vec.begin()+i);
}

File: CribbageFlutter/cribbage/lib/models/deck/card.dart
enum Suit { clubs, diamonds, hearts, spades }
enum Value {
  ace,
  two,
  three,
  four,
  five,
  six,
  seven,
  eight,
  nine,
  ten,
  jack,
  queen,
  king
}

class Card {
  final Value _value;
  final Suit _suit;

  Card(this._value, this._suit);

  bool operator <(Card card) {
    return _value.index < card._value.index;
  }

  int pips() {
    switch (_value) {
      case Value.ace:
        return 1;
      case Value.two:
        return 2;
      case Value.three:
        return 3;
      case Value.four:
        return 4;
      case Value.five:
        return 5;
      case Value.six:
        return 6;
      case Value.seven:
        return 7;
      case Value.eight:
        return 8;
      case Value.nine:
        return 9;
      case Value.ten:
        return 10;
      case Value.jack:
        return 10;
      case Value.queen:
        return 10;
      case Value.king:
        return 10;
    }
  }

  String valueStr() {
    switch (_value) {
      case Value.ace:
        return "ace";
      case Value.two:
        return "two";
      case Value.three:
        return "three";
      case Value.four:
        return "four";
      case Value.five:
        return "five";
      case Value.six:
        return "six";
      case Value.seven:
        return "seven";
      case Value.eight:
        return "eight";
      case Value.nine:
        return "nine";
      case Value.ten:
        return "ten";
      case Value.jack:
        return "jack";
      case Value.queen:
        return "queen";
      case Value.king:
        return "king";
    }
  }

  String suitStr() {
    switch (_suit) {
      case Suit.clubs:
        return "clubs";
      case Suit.diamonds:
        return "diamonds";
      case Suit.hearts:
        return "hearts";
      case Suit.spades:
        return "spades";
    }
  }

  String toCompactString() {
    String ret = "";
    switch (_value) {
      case Value.ace:
        ret += "A";
        break;
      case Value.jack:
        ret += "J";
        break;
      case Value.queen:
        ret += "Q";
        break;
      case Value.king:
        ret += "K";
        break;
      default:
        ret += pips().toString();
        break;
    }
    ret += " ";
    switch (_suit) {
      case Suit.clubs:
        ret += ":\u200d";
        break;
      case Suit.diamonds:
        ret += "D";
        break;
      case Suit.hearts:
        ret += "H";
        break;
      case Suit.spades:
        ret += "S";
        break;
    }
    return ret;
  }

  String toPrettyString() {
    String ret = "";
    switch (_value) {
      case Value.ace:
        ret += "A";
        break;
      case Value.jack:
        ret += "J";
        break;
      case Value.queen:
        ret += "Q";
        break;
      case Value.king:
        ret += "K";
        break;
      default:
        ret += pips().toString();
        break;
    }
    ret += " ";
    switch (_suit) {
      case Suit.clubs:
        ret += "♣";
        break;
      case Suit.diamonds:
        ret += "♦";
        break;
      case Suit.hearts:
        ret += "♥";
        break;
      case Suit.spades:
        ret += "♠";
        break;
    }
    return ret;
  }

  @override
  String toString() {
    return valueStr() + " " + suitStr();
  }

  Value getValue() {
    return _value;
  }

  Suit getSuit() {
    return _suit;
  }
}
